### PR TITLE
Add rule for writing GDS files

### PR DESCRIPTION
**Pull request by lpawelcz**
Closes https://github.com/hdl/bazel_rules_hdl/issues/20, closes https://github.com/hdl/bazel_rules_hdl/issues/54.

This PR adds `gds_write` rule which allows writing GDS files. The rule have a set of inputs and dependencies which are:
* **Implemented RTL** - the result of `place_and_route` target,
* **KLayout** - delivered as Python module
* **Python script** - for writing GDS files with KLayout module
* **KLayout technology file**  - XML file generated from KLayout GUI version, based on data from PDKs.
* **KLayout LEF file** - technology LEF file of the PDK that includes all relevant information regarding metal layers, vias, and spacing requirements
* **Additional LEF files** -  List of hardened macro LEF view files.
* **Additional GDS files** - List of platform GDS files
* **Fill config** - JSON file with configuration for metal fill

Apart from the rules itself, PR:
* adds KLayout python module into `pip_requirements.txt`
* imports and modifies [def2stream.py](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/99e8d3e9b399c7703fab26bef2b5770d39a2f724/flow/util/def2stream.py) script from OpenROAD-flow-scripts
* modifies `place_and_route` rule to also return `SynthesisInfo` provider which is convenient for accessing top module name for the GDS write script.
* modifies `StandardCellInfo` provider to also store GDS files from PDKs
* adds `density_fill` stage after `detailed_routing`

### CODE REVIEW

**Comment by lpawelcz** [comment]
Hi @QuantamHD and @proppy, please take a look at this PR

File: gds_write/build_defs.bzl
Comment (proppy) [inline gds_write/build_defs.bzl:60]: curious if that could be done if def2stream instead?

File: gds_write/build_defs.bzl
Comment (proppy) [inline gds_write/build_defs.bzl:29]: curious why we need to copy the lef file? can't we just use them from their current location?

File: gds_write/build_defs.bzl
Comment (proppy) [inline gds_write/build_defs.bzl:69]: is this necessary to resolve the new `klayout` requirements?

File: gds_write/build_defs.bzl
Comment (proppy) [inline gds_write/build_defs.bzl:76]: curious why we need an intermediate script rather than `run_shell`?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:3]: need that you're able to use the klayout python module rather than the full executable!

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:17]: curious why we don't use openroad for that?
https://openroad.readthedocs.io/en/latest/main/src/fin/README.html

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:18]: what is `seal` for?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:83]: is there an issue on the klayout repo already opened for that?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:130]: leftovers?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:142]: wouldn't the `VIA_` also be child of the top_cell in the DEF?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:156]: if we only copy the top level, did we actually need to clear the orphan cells?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:163]: should we use a flag instead?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:208]: should we use exception instead? so that can exit earlier in the flow?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:206]: do we actually want to write if we have an error?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:1]: license header and docstring?

File: gds_write/BUILD
Comment (proppy) [inline gds_write/BUILD:20]: should we have some `py_test` too?

File: synthesis/tests/BUILD
Comment (proppy) [inline synthesis/tests/BUILD:52]: can we get those thru dependencies?

**Review by proppy — CHANGES_REQUESTED** [review]
Sorry for the delay,

Looks very good! I made a first round of comments with nits and questions about the implementation.

File: gds_write/build_defs.bzl
Comment (lpawelcz) [reply gds_write/build_defs.bzl]: Managed to simplify the rule and remove copying files

File: gds_write/build_defs.bzl
Comment (lpawelcz) [reply gds_write/build_defs.bzl]: Turns out we can just set `tools` attribute in `run_shell` action to point to `py_binary` with our GDS generation script and the environment would be correct then.

File: gds_write/build_defs.bzl
Comment (lpawelcz) [reply gds_write/build_defs.bzl]: Managed to change that to single `run_shell`

File: place_and_route/build_defs.bzl
Comment (lpawelcz) [reply place_and_route/build_defs.bzl:63]: It would be helpful if `place_and_route` rule would return SynthesisInfo provider with the name of the `top_module` from the synthesized design. This name is required in GDS write script, see e.g. [here](https://github.com/hdl/bazel_rules_hdl/pull/169/files/1b70039d8ff2dc0e741ba9a020a606b99d9dc023#diff-9a0852b1664f229eada1c6cf7973ba4e8f17edd051dfbe1b4b560bc7c52c3787R135)

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py]: We can do that. I added `density_fill` stage right after `detailed_routing`

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py]: Looks like this issue was solved in https://github.com/KLayout/klayout/issues/673. I modified the script and bazel flow to perform density fill after detailed routing and then write final DEF file which is read during GDS write

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py]: This part of the code was removed with whole density fill support and reimplemented in bazel flow right after detailed routing stage

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py]: Removed leftover comments

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py]: Removed this part of the code as it seems to be unnecessary.

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py:88]: I guess it is actually unnecessary. I don't see the point of removing the orphans from `main_layout` if we copy one specific cell tree from it and then use this cell through the rest of the flow. Thank you for pointing that out.

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py:111]: Raising exception is now placed before writing GDS.

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py]: replaced with `raise LookupError()`

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py]: For me it looks impossible, especially when we explicitly copy only the tree of the top cell to the new layout and check for orphans in it. I think we can remove this check.

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py]: Replaced with `--gds-allow-empty` flag

File: synthesis/tests/BUILD
Comment (lpawelcz) [reply synthesis/tests/BUILD]: I'm afraid that if we place all those files under single `dependencies` attribute we would have problems with distinguishing which file is which as they will be stored on a list and we would have to rely on file extensions.

File: gds_write/BUILD
Comment (lpawelcz) [reply gds_write/BUILD:20]: [Here](https://github.com/antmicro/bazel_rules_hdl/blob/gds-write/synthesis/tests/BUILD#L47-L53) we test the rule that uses this `py_binary`. If that is not enough I can add `py_test` but I think it would not be much different compared to the linked test. Please let me know what do you think.

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py:33]: I'm not sure if I fully understand you here. Could you please explain what do you mean? Is there anything to change regarding that import or KLayout Python Module in general?

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py:1]: Added license and short docstring

**Comment by lpawelcz** [comment]
@proppy I responded to all of your comments. Please take another look at this PR

File: gds_write/build_defs.bzl
Comment (proppy) [inline gds_write/build_defs.bzl:66]: do we need this to be an attribute?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:5]: is that the original license from which you created the derivative? There might be some implication since bazel_rules_hdl is Apache 2.0, see: https://github.com/hdl/bazel_rules_hdl/blob/main/LICENSE

@QuantamHD any insights?

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:33]: should we refer to the origin of the script?

File: gds_write/def2stream.py
Comment (proppy) [reply gds_write/def2stream.py:40]: s/need/neat/, sorry for the confusion :) I'm glad you're using the python module here.

File: gds_write/def2stream.py
Comment (proppy) [inline gds_write/def2stream.py:113]: any reason we don't raise the exception in place?

File: place_and_route/build_defs.bzl
Comment (proppy) [reply place_and_route/build_defs.bzl:63]: /cc @QuantamHD

File: place_and_route/private/detailed_routing.bzl
Comment (proppy) [inline place_and_route/private/detailed_routing.bzl:58]: can you comment about why this change is required?

File: synthesis/tests/klayout.lyt
Comment (proppy) [inline synthesis/tests/klayout.lyt:1]: is there a way we can get this from the pdk deps?

File: synthesis/tests/BUILD
Comment (proppy) [reply synthesis/tests/BUILD]: see https://github.com/hdl/bazel_rules_hdl/tree/main/dependency_support/com_google_skywater_pdk

**Review by proppy — APPROVED** [review]
Nice work, just some nits and questions left.

I'll let @QuantamHD a chance to review before hitting the merge button.

File: gds_write/build_defs.bzl
Comment (lpawelcz) [reply gds_write/build_defs.bzl:66]: This is used to create implicit dependency on def2stream py_binary. `_gds_write` is a private attribute so it is not possible to override its default value. This way we can also refer to the script through `ctx.executable.<_attr_name>`.

I saw a similar solution in [STA rule](https://github.com/hdl/bazel_rules_hdl/blob/main/static_timing/build_defs.bzl#L78-L82) and [Verilator rule](https://github.com/hdl/bazel_rules_hdl/blob/main/verilator/defs.bzl#L209-L214) and I wanted to follow this pattern here as well.

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py:5]: This license is taken from [OpenROAD-flow-scripts](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/LICENSE_BUILD_RUN_SCRIPTS) but it was never a part of the [source file](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/util/def2stream.py) that I imported. However, the License File states that it applies to all build and run scripts from OpenROAD-flow-scripts so it applies also to `def2stream.py`.

As far as I know, it is possible to import BSD 3-licensed source files into Apache 2.0 project but it is required to keep the original license header in the imported file.

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py:33]: Added a note on the origins with link to repo, import date, git hash and link to the source file at mentioned revision.

File: gds_write/def2stream.py
Comment (lpawelcz) [reply gds_write/def2stream.py:119]: We can do that but I wanted to let the script print out all the missing cells first before raising the exception.
Let me know if you would prefer to have this script fail early.

File: synthesis/tests/klayout.lyt
Comment (lpawelcz) [reply synthesis/tests/klayout.lyt:1]: In skywater-pdk repositories there are no KLayout files. Commit messages in https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/299 indicate that KLayout files were taken from [RTimothyEdwards/open_pdks](https://github.com/RTimothyEdwards/open_pdks) and those in open_pdks are taken from [mabrains/sky130_klayout_pdk](https://github.com/mabrains/sky130_klayout_pdk).

We could possibly add  `mabrains/sky130_klayout_pdk` here as dependency and write some new rules for fetching this file. Please let me know what do you think about that.

File: gds_write/build_defs.bzl
Comment (QuantamHD) [reply gds_write/build_defs.bzl:66]: LGTM.

File: place_and_route/build_defs.bzl
Comment (QuantamHD) [reply place_and_route/build_defs.bzl:63]: Why not just return the ctx.attr.synthesized_rtl[SynthesisInfo] directly there's really no harm.

File: synthesis/tests/sky130_fd_sc_hd.gds
Comment (QuantamHD) [inline synthesis/tests/sky130_fd_sc_hd.gds:1]: The GDS and LEF should be available in the sky130 dependency. Ideally, we would add them to the SynthesisInfo's StandardCellInfo provider https://github.com/hdl/bazel_rules_hdl/blob/main/pdk/build_defs.bzl#L18

The raw files are collected by this file:
https://github.com/hdl/bazel_rules_hdl/blob/main/dependency_support/com_google_skywater_pdk/declare_cell_library.bzl

I'll approve without this though since this is just a test, but the idea of the bazel rules is that you only need to specify the PDK files in the synthesis rule, and all other files are transitively available to the subsequent rules. Including info like GDS and LEF

**Comment by QuantamHD** [comment]
Sorry for the long poll in review time. I'm committed to get that down to 24 hours. I will be more responsive going forward.

File: place_and_route/build_defs.bzl
Comment (lpawelcz) [reply place_and_route/build_defs.bzl:63]: Done

File: synthesis/tests/BUILD
Comment (lpawelcz) [reply synthesis/tests/BUILD]: Oh, I see. I was able to use PDK dependencies to pass down LEF and GDS files through `StandardCellInfo` provider. From now on there is no need to specify `input_lef` and `input_gds` attributes. However I've left those attributes - they might be useful in some custom cases.

File: synthesis/tests/sky130_fd_sc_hd.gds
Comment (lpawelcz) [reply synthesis/tests/sky130_fd_sc_hd.gds:1]: Thank you, that was really helpful. With your hints I was able to use LEF and GDS files from sky130 dependency (as described in https://github.com/hdl/bazel_rules_hdl/pull/169#discussion_r1258214673). I've also attempted to modify `StandardCellInfo` for `ASAP7` but I'm not sure if this change should land in this PR

File: synthesis/tests/klayout.lyt
Comment (lpawelcz) [reply synthesis/tests/klayout.lyt:1]: Klayout *.lyt file has 3 different sources from which it could be imported to [OpenROAD-flow-scripts](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/platforms/sky130hd/sky130hd.lyt):
* [mabrains/sky130_klayout_pdk](https://github.com/mabrains/sky130_klayout_pdk/blob/main/sky130_tech/tech/sky130/sky130.lyt)
* [efabless/sky130_klayout_pdk](https://github.com/efabless/sky130_klayout_pdk/blob/main/sky130_tech/tech/sky130/sky130.lyt)
* [RTimothyEdwards/open_pdks](https://github.com/RTimothyEdwards/open_pdks/blob/master/sky130/klayout/sky130.lyt)

I tested all of them and compared against *.lyt from `ORFS` and it looks like the most similar is the file from `open_pdks`. There are also some differences in GDS files generated with those KLayout configurations. The ones generated from configurations taken from `ORFS` and `open_pdks` look alright but those generated with configurations from `efabless` and `mabrains` do not have any paths between cells on layers `met2`, `met3`, `met4`

So `open_pdks` looks like a good candidate for fetching those files from dependencies. But there is a catch: open_pdks does not support ASAP7 and doesn't store any files for this technology - ORFS does. If we want to get `fill_config.json` and `klayout.lyt` from dependencies then I think ORFS would be the best repository for fetching this kind of data.

**Review by QuantamHD — APPROVED** [review]

**Comment by QuantamHD** [comment]
Thanks!

**Pull request merged**
